### PR TITLE
Pr92x L1T uGT emulation M_T and diObjectPT

### PR DESCRIPTION
**Pull request by rekovic**
pr92x L1T uGT emulation M_T and diObjectPT.

This had to wait for the UTM 0.6.3 to be merged, now part of CMSSW_9_2_1. 

Needed for data taking, by ECAL AlCa paths that seed with L1T M_T algorithms.

### CODE REVIEW

**Comment by rekovic** [comment]
please test

**Comment by rekovic** [comment]
+1

**Comment by ggovi** [comment]
@rekovic 
Hi Vladimir, unfortunately these changes can't be done. The CondFormats classes used in the persistency cannot be extended, since the older releases won't be able to read the new data.
You can only add data which you don't need to store in the DB, by declare them COND_TRANSIENT

**Comment by rekovic** [comment]
@ggovi 
Thanks.

**Comment by rekovic** [comment]
@ggovi 
I looked again (and remembered) that  the changes introduced in files in CondFormats 

<pre>
CondFormats/L1TObjects/interface/L1TUtmAlgorithm.h
CondFormats/L1TObjects/interface/L1TUtmBin.h
CondFormats/L1TObjects/interface/L1TUtmCondition.h
CondFormats/L1TObjects/interface/L1TUtmCut.h
CondFormats/L1TObjects/interface/L1TUtmCutValue.h
CondFormats/L1TObjects/interface/L1TUtmObject.h
CondFormats/L1TObjects/interface/L1TUtmScale.h
CondFormats/L1TObjects/interface/L1TUtmTriggerMenu.h
</pre>

 there are no new data members which are added to the classes, 
only setters and getters are introduced. 
This should not affect the backward compatibility, as far as I understand.

Thanks

**Comment by ggovi** [comment]
@rekovic 
sorry, I've overlooked this thing. It's ok then.

**Comment by ggovi** [comment]
+1

**Comment by rekovic** [comment]
@arunhep, @franzoni  ping

**Comment by arunhep** [comment]
+1

File: L1Trigger/L1TGlobal/interface/CorrelationTemplate.h [hunk 1]
```diff
@@ -66,25 +66,29 @@ class CorrelationTemplate : public GlobalCondition
...
 
 	long long minDRCutValue;
 	long long maxDRCutValue;
-	unsigned int precDRCut; 
+	unsigned int precDRCut;
 
 	long long minMassCutValue;
 	long long maxMassCutValue;
-	unsigned int precMassCut; 
+	unsigned int precMassCut;
```
Comment (davidlange6) [inline L1Trigger/L1TGlobal/interface/CorrelationTemplate.h]: hi @rekovic - please fix the indentation here.

File: L1Trigger/L1TGlobal/plugins/TriggerMenuParser.cc [hunk 2]
```diff
@@ -974,14 +997,59 @@ void l1t::TriggerMenuParser::parseDeltaPhi_Cos_LUTS(std::map<std::string, tmeven
...
     // Second Get the Cosh for this delta phi Set
     std::vector<long long> lut_cos;
     applyCos(val_delta_phi, n);
     setLut(lut_cos, val_delta_phi, prec2);
     m_gtScales.setLUT_Cos(lutName,lut_cos,prec2);
 
-}	
+}
+
+void l1t::TriggerMenuParser::parsePhi_Trig_LUTS(std::map<std::string, tmeventsetup::esScale> scaleMap, std::string obj, std::string func, unsigned int prec)
```
Comment (davidlange6) [inline L1Trigger/L1TGlobal/plugins/TriggerMenuParser.cc]: please pass things as const &

File: L1Trigger/L1TGlobal/plugins/TriggerMenuParser.cc [hunk 3]
```diff
@@ -974,14 +997,59 @@ void l1t::TriggerMenuParser::parseDeltaPhi_Cos_LUTS(std::map<std::string, tmeven
...
+{
+
+    using namespace tmeventsetup;
+
+    std::string scLabel = obj;
+    scLabel += "-PHI";
+
+    //This LUT does not exist in L1 Menu file, don't fill it
+    if( scaleMap.find(scLabel) == scaleMap.end() ) return;
+    if( func != "Sin" and func != "Cos") return;
```
Comment (davidlange6) [inline L1Trigger/L1TGlobal/plugins/TriggerMenuParser.cc]: could fund be an enum?

File: L1Trigger/L1TGlobal/plugins/TriggerMenuParser.cc [hunk 4]
```diff
@@ -974,14 +997,59 @@ void l1t::TriggerMenuParser::parseDeltaPhi_Cos_LUTS(std::map<std::string, tmeven
...
+
+    const esScale* scale = &scaleMap.find(scLabel)->second;
+
+    const double step = scale->getStep();
+    const double range = scale->getMaximum() - scale->getMinimum();
+    const size_t n = std::ceil(range / step);
+    const size_t bitwidth = std::ceil(std::log10(n) / std::log10(2));
+
+    std::vector<double> array;
```
Comment (davidlange6) [inline L1Trigger/L1TGlobal/plugins/TriggerMenuParser.cc]: these three lines can be more efficiently written as
std::vector<double> array(std::pow(2,bitwidth),0)

File: L1Trigger/L1TGlobal/plugins/TriggerMenuParser.cc (hunk 2 above)
Comment (davidlange6) [inline L1Trigger/L1TGlobal/plugins/TriggerMenuParser.cc]: looks like there are only two valid values.

File: L1Trigger/L1TGlobal/src/GlobalScales.cc [hunk 5]
```diff
@@ -148,28 +148,44 @@ void l1t::GlobalScales::setLUT_Cos(std::string lutName, std::vector<long long> l
 
 }
 
+void l1t::GlobalScales::setLUT_Sin(std::string lutName, std::vector<long long> lut, unsigned int precision)
```
Comment (davidlange6) [inline L1Trigger/L1TGlobal/src/GlobalScales.cc]: pass as const &

File: L1Trigger/L1TGlobal/src/GlobalScales.cc [hunk 6]
```diff
@@ -235,219 +251,284 @@ long long l1t::GlobalScales::getLUT_DeltaPhi(std::string lutName, int element) c
...
+
+   if(element < (int)m_lut_Cos.find(lutName)->second.size()) {
+     value = m_lut_Cos.find(lutName)->second.at(element);
+   } else {
+     edm::LogError("GlobalScales") << "Warning: Element Requested " << element << " too large for Cos LUT (" << lutName <<") size = " << m_lut_Cos.find(lutName)->second.size() << std::endl;
+   }
+   return value;
+}
+
+long long l1t::GlobalScales::getLUT_Sin(std::string lutName, int element) const
```
Comment (davidlange6) [inline L1Trigger/L1TGlobal/src/GlobalScales.cc]: similarly here.

**Comment by davidlange6** [comment]
and generically - i don't understand how "version_" is going to be changed from something other than 0 in the future - please explain

**Comment by davidlange6** [comment]
-1
pending answers to comments

File: L1Trigger/L1TGlobal/interface/CorrelationTemplate.h (hunk 1 above)
Comment (rekovic) [inline L1Trigger/L1TGlobal/interface/CorrelationTemplate.h]: done.

File: L1Trigger/L1TGlobal/plugins/TriggerMenuParser.cc (hunk 2 above)
Comment (rekovic) [inline L1Trigger/L1TGlobal/plugins/TriggerMenuParser.cc]: done.

yes, only two valid values "sin" and "cos", now passed as const &. 
Will request developer to use enums in the next version.  I hope OK for now.

File: L1Trigger/L1TGlobal/plugins/TriggerMenuParser.cc (hunk 3 above)
Comment (rekovic) [inline L1Trigger/L1TGlobal/plugins/TriggerMenuParser.cc]: Will request developer to use enums in the next version.  I hope OK for now.

File: L1Trigger/L1TGlobal/plugins/TriggerMenuParser.cc (hunk 4 above)
Comment (rekovic) [inline L1Trigger/L1TGlobal/plugins/TriggerMenuParser.cc]: Done.

File: L1Trigger/L1TGlobal/src/GlobalScales.cc (hunk 5 above)
Comment (rekovic) [inline L1Trigger/L1TGlobal/src/GlobalScales.cc]: done

**Comment by rekovic** [comment]
@davidlange6 
> and generically - i don't understand how "version_" is going to be changed from something other than 0 in the future - please explain

You are right, there are no setters of version, apart from what is done in constructor.  
Added setters in the constructors like "version(0)" are added for already existing member data. 
which were not set in the previous version, as for example here
https://github.com/cms-sw/cmssw/pull/19155/files#diff-c571023383d9c2a73069f92875823735R42
Setting it to "0" seems ok to me.

**Comment by rekovic** [comment]
please test

**Comment by rekovic** [comment]
@davidlange6 
In last commit, used enum instead of string as arguments to member functions, as requested.

**Comment by rekovic** [comment]
please test

**Comment by davidlange6** [comment]
merge

**Pull request merged**